class-two-factor-core.php

```diff
@@ -731,12 +731,20 @@ public static function login_url( $params = array(), $scheme = 'login' ) {
 	/**
 	 * Get the hash of a nonce for storage and comparison.
 	 *
-	 * @param string $nonce Nonce value to be hashed.
+	 * @param array $nonce Nonce array to be hashed. ⚠️ This must contain user ID and expiration,
+	 *                     to guarantee the nonce only works for the intended user during the
+	 *                     intended time window.
 	 *
-	 * @return string
+	 * @return string|false
 	 */
 	protected static function hash_login_nonce( $nonce ) {
-		return wp_hash( $nonce, 'nonce' );
+		$message = wp_json_encode( $nonce );
+
+		if ( ! $message ) {
+			return false;
+		}
+
+		return hash_hmac( 'sha256', $message, wp_salt( 'nonce' ) );
```

Comment:
What is is the PHP support for `hash_hmac` with this specific hash function? Is this a breaking change? What are the benefits compared to using `wp_hash()` helper for this?

Reply:
[It works all the way down to PHP 5.2](https://3v4l.org/Wr8c4#v5.2.0)

Reply:
Not quite, the WordPress included `hash_hmac()` fallback only supports `md5` and `sha1`.

https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/compat.php#L253-L271

For this reason, core uses sha256 and falls back to sha1 in auth:

https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/pluggable.php#L761-L767

The compat is still included as [the hash extension can be disabled until PHP 7.4](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/compat.php#L228-L233).

Reply:
Should we stick to `wp_hash()` here since the general idea is to (1) _not_ have the token in plain text and (2) have it work on as many systems as possible? These are short lived tokens and it is seems to be more important that they get hashed than which hash function is used. Before this change we used `wp_hash()` which "guaranteed" that it will be hashed as expected.

The addition of user ID and the timestamp to the hashed payload is the important part here, right?

Reply:
Ah, great point Dion 👍🏻 

🤔 , I definitely prefer using Core functions in general, this just seemed like an exception given how insecure MD5 is.  Now that I read about it more, though, it does seem like it's safe-ish in an HMAC context (if I'm understanding everything correctly).

* https://crypto.stackexchange.com/a/9340/9297
* https://crypto.stackexchange.com/a/6753/9297 - in our case the `NONCE_SALT` key length is only 64 chars, but [that's still impractical against a remote network](https://www.wolframalpha.com/input?i=2%5E63+%2F+150000)

I'm still leaning towards erring on the side of caution, though; especially given how easy it is to use [Core's progressive enhancement technique](https://github.com/WordPress/wordpress-develop/blob/3977b6b06d1efd3f6cadb8b31bc8ba55e09486d5/src/wp-includes/pluggable.php#L764). That's partially just because I'm far from an expert here, and I'd rather be safe than sorry.

Reply:
You can also use sodium_crypto_generic_hash

Reply:
The primary reason the Auth cookies above use `hash_hmac()` over `wp_hash()` is that they need to use a custom `$key` that changes per-user in addition to per-site (which is why the key also uses `wp_hash()` in it), rather than `wp_hash()` which just uses the static per-site key.

In this case, `wp_hash()` (which uses `hash_mac( 'md5' )`) is secure IMO for this purposes. I don't think this is a scenario where `md5 is insecure so we shouldn't use it` holds true.

Reply:
That works for me, I moved it back to `wp_hash()` in [b153455](https://github.com/WordPress/two-factor/pull/473/commits/b153455dfab0ee7b2624d44e9c5510e783b47dd9)